### PR TITLE
#15/Add go conversation to pkg/convert/go_messages.go

### PR DESCRIPTION
**Pull request by DarjaLeonova**

### CODE REVIEW

File: pkg/convert/go_messages.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+func EncodeToBytes(m model.Messages) ([]byte, error) {
+	message := model.Messages{
+		Messages: m.Messages,
+		Language: m.Language,
+		Labels:   m.Labels,
+	}
+
+	msg, err := json.Marshal(message)
+	if err != nil {
+		fmt.Println(fmt.Errorf("error while marshaling: %w", err))
```
Comment (VladislavsPerkanuks) [inline pkg/convert/go_messages.go]: Return empty bytes and error

File: pkg/convert/go_messages.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+	return msg, nil
+}
+
+func DecodeToMessages(b []byte) model.Messages {
+	var msg model.Messages
+
+	err := json.Unmarshal(b, &msg)
+	if err != nil {
+		fmt.Println(fmt.Errorf("error while unmarshaling: %w", err))
```
Comment (VladislavsPerkanuks) [inline pkg/convert/go_messages.go]: Return empty model and error

**Review by VladislavsPerkanuks — CHANGES_REQUESTED** [review]

File: pkg/convert/go_messages.go [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/expect-digital/translate/pkg/model"
+)
+
+func FromGoMessages(m model.Messages) ([]byte, error) {
+	msg, err := json.Marshal(m)
```
Comment (janishorsts) [inline pkg/convert/go_messages.go]: We require to transform `model.Messages` into `pipelines.Messages` before JSON encoding. `model.Messages`.

Comment (DarjaLeonova) [reply pkg/convert/go_messages.go]: Aaaa I get it ;D

Comment (DarjaLeonova) [reply pkg/convert/go_messages.go]: I think I done, but not really sure. Check when you will have time ;)

File: pkg/convert/go_messages.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+	err := json.Unmarshal(b, &msg)
+	if err != nil {
+		return pipeline.Messages{}, fmt.Errorf("error while unmarshaling: %w", err)
+	}
+
+	return msg, nil
+}
+
+func MessagesToPipeline(m model.Messages) pipeline.Messages {
```
Comment (janishorsts) [inline pkg/convert/go_messages.go]: We don't want to expose this function.

```suggestion
func messagesToPipeline(m model.Messages) pipeline.Messages {
```

Comment (janishorsts) [inline pkg/convert/go_messages.go]: We don't want to expose this function.

```suggestion
func messagesToPipeline(m model.Messages) pipeline.Messages {
```

File: pkg/convert/go_messages.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+	pipelineMsg := pipeline.Messages{}
```
Comment (janishorsts) [inline pkg/convert/go_messages.go]: Declare and initialize all values in one go

```suggestion
	pipelineMsg := pipeline.Messages{Language: m.Language}
```

Comment (janishorsts) [inline pkg/convert/go_messages.go]: Declare and initialize all values in one go

```suggestion
	pipelineMsg := pipeline.Messages{Language: m.Language, Messages: make([]pipeline.Message, 0, len(m.Messages)}
```

File: pkg/convert/go_messages.go [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+	pipelineMsg.Language = m.Language
+
+	for key, value := range m.Labels {
+		pipelineMsg.Macros[key] = pipeline.Text{Msg: value}
+	}
```
Comment (janishorsts) [inline pkg/convert/go_messages.go]: Delete this one. We use Labels only internally.

```suggestion
```

Comment (janishorsts) [inline pkg/convert/go_messages.go]: Delete this one. We use Labels only internally.

```suggestion
```

Comment (janishorsts) [inline pkg/convert/go_messages.go]: Delete this one. We use Labels only internally.

```suggestion
```

File: pkg/convert/go_messages.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+	"github.com/expect-digital/translate/pkg/model"
+	"golang.org/x/text/message/pipeline"
+)
+
+func FromGoMessages(m model.Messages) ([]byte, error) {
+	pipelineMsg := MessagesToPipeline(m)
+
+	msg, err := json.Marshal(pipelineMsg)
+	if err != nil {
+		return []byte{}, fmt.Errorf("error while marshaling: %w", err)
```
Comment (janishorsts) [inline pkg/convert/go_messages.go]:
```suggestion
		return nil, fmt.Errorf("error while marshaling: %w", err)
```

File: pkg/convert/go_messages.go [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+	}
+
+	return msg, nil
+}
+
+func ToGoMessages(b []byte) (pipeline.Messages, error) {
```
Comment (janishorsts) [inline pkg/convert/go_messages.go]:
```suggestion
func ToGoMessages(b []byte) (model.Messages, error) {
```

**Review by janishorsts — CHANGES_REQUESTED** [review]

File: go.mod [hunk 9]
```diff
@@ -12,9 +12,11 @@ require (
 	github.com/golang/protobuf v1.5.2 // indirect
 	github.com/pmezard/go-difflib v1.0.0 // indirect
 	github.com/stretchr/testify v1.8.1 // indirect
+	golang.org/x/mod v0.6.0-dev.0.20220419223038-86c51ed26bb4 // indirect
 	golang.org/x/net v0.0.0-20220722155237-a158d28d115b // indirect
 	golang.org/x/sys v0.0.0-20220722155257-8c9f86f7a55f // indirect
-	golang.org/x/text v0.4.0 // indirect
+	golang.org/x/text v0.5.0 // indirect
```
Comment (janishorsts) [inline go.mod]: `golang.org/x/text` is direct dependency. Please run, `go mod tidy`.

File: pkg/convert/go_messages.go [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+
+func messagesToPipeline(m model.Messages) pipeline.Messages {
+	pipelineMsg := pipeline.Messages{
+		Language: m.Language,
+		Messages: make([]pipeline.Message, 0, len(m.Messages)),
+	}
+
+	for _, value := range m.Messages {
+		pipelineMsg.Messages = append(pipelineMsg.Messages, pipeline.Message{
+			ID:      append(pipeline.IDList{}, value.ID),
```
Comment (janishorsts) [inline pkg/convert/go_messages.go]:
```suggestion
			ID:      pipeline.IDList{value.ID},
```

File: pkg/convert/go_messages.go [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+		return nil, fmt.Errorf("error while marshaling: %w", err)
+	}
+
+	return msg, nil
+}
+
+func ToGoMessages(b []byte) (model.Messages, error) {
+	var msg model.Messages
+
+	err := json.Unmarshal(b, &msg)
```
Comment (janishorsts) [inline pkg/convert/go_messages.go]: `b` contains JSON encoded `pipeline.Messages`, not `model.Messages`.

**Review by janishorsts — CHANGES_REQUESTED** [review]

File: pkg/convert/go_messages.go [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+	"github.com/expect-digital/translate/pkg/model"
+	"golang.org/x/text/message/pipeline"
+)
+
+func FromGoMessages(m model.Messages) ([]byte, error) {
+	pipelineMsg := messagesToPipeline(m)
+
+	msg, err := json.Marshal(pipelineMsg)
+	if err != nil {
+		return nil, fmt.Errorf("error while marshaling: %w", err)
```
Comment (VladislavsPerkanuks) [inline pkg/convert/go_messages.go]: "error" is redundant. We know that this is error, better explain what you were trying to accomplish, e.g.  "marshal to go messages"

File: pkg/convert/go_messages.go [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+
+	return msg, nil
+}
+
+func ToGoMessages(b []byte) (model.Messages, error) {
+	var pipelineMsg pipeline.Messages
+
+	err := json.Unmarshal(b, &pipelineMsg)
+	if err != nil {
+		return model.Messages{}, fmt.Errorf("error while unmarshaling: %w", err)
```
Comment (VladislavsPerkanuks) [inline pkg/convert/go_messages.go]: "error" is redundant.

File: pkg/convert/go_messages.go [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+	"github.com/expect-digital/translate/pkg/model"
+	"golang.org/x/text/message/pipeline"
+)
+
+func FromGoMessages(m model.Messages) ([]byte, error) {
+	pipelineMsg := messagesToPipeline(m)
+
+	msg, err := json.Marshal(pipelineMsg)
+	if err != nil {
+		return nil, fmt.Errorf("failed pipeline.Messages marshaling: %w", err)
```
Comment (janishorsts) [inline pkg/convert/go_messages.go]: Advisory. Describe the intent when wrapping error messages. No need to emphasize something failed. It's an error message, we know something failed. It makes it more succinct and clearer. E.g. `encode pipeline.Messages to JSON: %w`.

File: pkg/convert/go_messages.go [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+
+func messagesFromPipeline(m pipeline.Messages) model.Messages {
+	msg := model.Messages{
+		Language: m.Language,
+		Messages: make([]model.Message, 0, len(m.Messages)),
+	}
+
+	for i, value := range m.Messages {
+		msg.Messages = append(msg.Messages, model.Message{
+			ID:      value.ID[i],
```
Comment (janishorsts) [inline pkg/convert/go_messages.go]: Wrong use of `i`. It's gonna panic.  Use first id value, and ignore remaining ones.

```suggestion
			ID:      value.ID[0], // use the first identifier and ignore remaining ones
```

**Review by janishorsts — CHANGES_REQUESTED** [review]

**Review by janishorsts — APPROVED** [review]

**Pull request merged**